include/picongpu/particles/debyeLength/Check.hpp

```diff
@@ -0,0 +1,180 @@
+                 *
+                 * This function must be called from all MPI ranks.
+                 *
+                 * @tparam T_ElectronSpecies electron species type
+                 *
+                 * @param cellDescription mapping for kernels
+                 */
+                template<typename T_ElectronSpecies>
+                HINLINE Status checkDebyeLength(MappingDesc const cellDescription)
+                {
+                    auto localResult = checkLocalDebyeLength<T_ElectronSpecies>(cellDescription);
+                    auto globalResult = Result{};
+                    pmacc::mpi::MPIReduce reduce;
+                    reduce(
+                        pmacc::nvidia::functors::Add(),
+                        &globalResult.numViolatingSupercells,
+                        &localResult.numViolatingSupercells,
+                        1,
+                        pmacc::mpi::reduceMethods::AllReduce());
+                    reduce(
+                        pmacc::nvidia::functors::Add(),
+                        &globalResult.numActiveSupercells,
+                        &localResult.numActiveSupercells,
+                        1,
+                        pmacc::mpi::reduceMethods::AllReduce());
+                    if(globalResult.numActiveSupercells)
+                    {
+                        auto const ratioFailing = static_cast<float_64>(globalResult.numViolatingSupercells)
+                            / static_cast<float_64>(globalResult.numActiveSupercells);
+                        /* Due to random nature of temperature and estimates used internally,
+                         * we allow a relatively small number of supercells to fail the check
+                         */
+                        auto const thresholdFailing = 0.1;
+                        if(ratioFailing <= thresholdFailing)
+                            return Status::Passed;
+                        else
+                            return Status::Failed;
+                    }
+                    else
+                        return Status::Skipped;
+                }
+
+            } // namespace detail
+
+            /** Check Debye length resolution
+             *
+             * The check is currently done only for species called "e" and is supposed
+             * to be called just after the particles are initialized as start of a
+             * simulation. The results are output to log< picLog::PHYSICS >.
+             *
+             * This function must be called from all MPI ranks.
+             *
+             * @param cellDescription mapping for kernels
+             */
+            HINLINE void check(MappingDesc const cellDescription)
+            {
+                bool isPrinting = (Environment<simDim>::get().GridController().getGlobalRank() == 0);
+                using ElectronSpecies = pmacc::particles::meta::FindByNameOrType_t<
+                    VectorAllSpecies,
+                    PMACC_CSTRING("e"),
```

Comment:
We need to implement another way to find all-electron species because there is no guarantee that this is the name for electrons.
You suggested already to check the charge of each species. The fix should be one helper function only, take care you can have more than one electron species.

Reply:
A possible way is to have an entry function for the check and that you call this entry function for each species.
Add a guard that only electrons can pass.

```C++
template<typename T_Species)
void check(T_Species sp)
{
   if(chargeOf<T_Species>!= ELECTRON_CHARGE)
       return;
   //...
}
```

Reply:
Yes, I thought about it. But then I realized that if there are multiple electron species, there is no easy way to do this check at all. So I decided to only implement the understandable and common case of one electron species.

Reply:
Simply call this check for all species and do a return if the species is not an electron species.

Reply:
You can call a functor for each species with `meta::ForEach` e.g. https://github.com/ComputationalRadiationPhysics/picongpu/blob/84e03980f2a56c7aea24d88bc3be9eb43f1a3197/include/picongpu/plugins/openPMD/openPMDWriter.hpp#L266

Reply:
offline discussed: we will check how many electrons like species with a current solver are used. In case only one electron species is used the test will be performed, else there will be a warning that this test is not executed.